src/contracts/CamelotYakRouter.sol

```diff
@@ -397,24 +397,27 @@ contract CamelotYakRouter is Maintainable, Recoverable, IYakRouter {
             amountIn = _applyFee(_trade.amountIn, _fee);
             _transferFrom(_trade.path[0], _from, FEE_CLAIMER, _trade.amountIn - amountIn);
         }
-        _transferFrom(_trade.path[0], _from, _trade.adapters[0], amountIn);
+        uint256 recipientBalanceBefore = IERC20(_trade.path[0]).balanceOf(_trade.recipients[0]);
+        _transferFrom(_trade.path[0], _from, _trade.recipients[0], amountIn);
+        amountIn = IERC20(_trade.path[0]).balanceOf(_trade.recipients[0]) - recipientBalanceBefore;
 
         address tokenOut = _trade.path[_trade.path.length - 1];
-        uint256 balanceBefore = IERC20(tokenOut).balanceOf(_to);
         for (uint256 i = 0; i < _trade.adapters.length; i++) {
             // All adapters should transfer output token to the following target
             // All targets are the adapters, expect for the last swap where tokens are sent out
-            address targetAddress = i < _trade.adapters.length - 1 ? _trade.adapters[i + 1] : _to;
+            address targetAddress = i < _trade.adapters.length - 1 ? _trade.recipients[i + 1] : _to;
+
+            recipientBalanceBefore =  IERC20(_trade.path[i + 1]).balanceOf(targetAddress);
             IAdapter(_trade.adapters[i]).swap(
                 amountIn,
                 0,
                 _trade.path[i],
                 _trade.path[i + 1],
                 targetAddress
             );
-            amountIn = IERC20(_trade.path[i + 1]).balanceOf(targetAddress);
+            amountIn = IERC20(_trade.path[i + 1]).balanceOf(targetAddress) - recipientBalanceBefore;
         }
-        uint256 amountOut = amountIn - balanceBefore;
+        uint256 amountOut = amountIn;
```

Comment:
You could remove this variable assignment now since you can use `amountIn` directly

Reply:
you're right, we actually discussed it with @0xpercival yesterday, we ended up keeping like this because even if slightly less optimized it feels more clear